### PR TITLE
Minor: Remove unnecessary name field in `ScalarFunctionDefintion`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to / follow on to  https://github.com/apache/arrow-datafusion/pull/8258  from @2010YOUY01 

## Rationale for this change

While integrating https://github.com/apache/arrow-datafusion/pull/8258 into our downstream system, I found creating a `BuildInScalarFunction` to be quite awkward as it had a seemingly duplicated name field. I thought the name field was to handle aliases or something, but it turns out I could remove it easily. 

## What changes are included in this PR?

Basically change the `ScalarFunctionDefinition::BuiltIn` to match the others

```rust
    BuiltIn {
        fun: built_in_function::BuiltinScalarFunction,
        name: Arc<str>,
    },
```

to 

```rust
    BuiltIn(BuiltinScalarFunction),
```
## Are these changes tested?
Covered by existing tests

## Are there any user-facing changes?

While this is technically an API change the API hasn't been released so it isn't an additional breaking API change

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -342,10 +342,7 @@ pub enum ScalarFunctionDefinition {
     /// Resolved to a `BuiltinScalarFunction`
     /// There is plan to migrate `BuiltinScalarFunction` to UDF-based implementation (issue#8045)
     /// This variant is planned to be removed in long term
-    BuiltIn {
-        fun: built_in_function::BuiltinScalarFunction,
-        name: Arc<str>,
-    },
+    BuiltIn(BuiltinScalarFunction),
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:345]: This is the core change here -- don't keep another string around, but just use the function on demand

**Review by viirya — APPROVED** [review]
Looks more clear to me.
